docs/src/proposals/comprehensive-compute-fees-review.md

```diff
@@ -0,0 +1,164 @@
+---
+title: Comprehensive Compute Fees
+---
+
+## Motivation
+
+The current fee structure lacks a comprehensive account of the work required by
+a validator to process a transaction.  The fee structure is only based on the
+number of signatures in a transaction but is meant to account for the work that
+the validator must perform to validate each transaction.  The validator performs
+a lot more user-defined work than just signature verification.  Processing a
+transaction typically includes signature verifications, account locking, account
+loading, and instruction processing.
+
+## Proposed Solution
+
+The following solution does not specify what native token costs are to be
+associated with the new fee structure.  Instead, it sets the criteria and
+provides the knobs that a cost model can use to determine those costs.
+
+### Fee
+
+The goal of the fees are to prevent the abuse of on-chain compute resources when
+processing  a transaction.  Each of the fee categories below will be represented
+as a compute unit cost that, when added together, encompasses the entire cost of
+processing the transaction.  By calculating the total cost of the transaction,
+the runtime can charge a more representative fee and make better transaction
+scheduling decisions.
+
+A fee will be calculated based on:
+
+1. Number of signatures
+   - Fixed rate per signature
+2. Number of write locks
```

Comment:
I'd like to suggest the following cost on write locks instead:
1. The bank keeps track of a new sysvar `WriteLock1111..`. The sysvar has the following data: `HashMap<Pubkey, DequeueArray<u32, N>>` where the dequeue represents the number of write locks acquired for each of the N previous blocks.
2. When calling `accounts.lock_accounts`, the bank keeps track of how many write locks occur on each account.
3. At the end of a slot, the bank commits the number of write locks for every account write locked to the corresponding dequeue in `WriteLock111...`. We update hashmap entries with no account writelocks with 0, and entries with `[0; N]` arrays are garbage collected.

Finally, we apply a cost model that is sufficiently coarse-grained to the total number of locks acquired, so that that the transaction cost does not vary too dynamically. It is fixed below a certain threshold, and increases at an increasing rate beyond that threshold. Finding the right threshold and rate of increase remains to be investigated.

The purpose of this approach is to tax high-contention accounts.

As we've seen from the sept liveness failure event, account contention is a real issue that can prevent parallelization to allow linearization of transaction processing. It would also make griefing more expensive.

Although an approach to offering more fine-grained parallelism in executing transactions could probably help as well, fundamentally speaking, we don't really want the same account to be write locked in every Entry.

Reply:
Interesting and thanks for taking the time to read the proposal.  How does this approach help the example given?  Without recoding read-locks, it is possible for a tx with a write-lock to be cheap but lock a normally read-only and heavily used account.  If we recorded all locks then a request to lock a heavily used account (read or write-locked) 
would be more expensive to write lock.